types/wlr_surface_layers.c

```diff
@@ -0,0 +1,212 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/types/wlr_surface.h>
+#include <wlr/types/wlr_surface_layers.h>
+#include <wlr/util/log.h>
+#include "surface-layers-protocol.h"
+
+static const char *surface_layers_role = "layer_surface";
+
+static void layer_surface_set_interactivity(struct wl_client *client,
+		struct wl_resource *resource, uint32_t input_types,
+		uint32_t exclusive_types) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->input_types = input_types;
+	surface->exclusive_types = exclusive_types;
+	wl_signal_emit(&surface->events.set_interactivity, surface);
+}
+
+static void layer_surface_set_anchor(struct wl_client *client,
+		struct wl_resource *resource, uint32_t anchor) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->anchor = anchor;
+	wl_signal_emit(&surface->events.set_anchor, surface);
+}
+
+static void layer_surface_set_exclusive_zone(struct wl_client *client,
+		struct wl_resource *resource, uint32_t zone) {
+	struct wlr_layer_surface *surface = wl_resource_get_user_data(resource);
+	surface->exclusive_zone = zone;
+	wl_signal_emit(&surface->events.set_exclusive_zone, surface);
+}
+
+static void layer_surface_set_margin(struct wl_client *client,
+		struct wl_resource *resource, uint32_t horizontal,
+		uint32_t vertical) {
```

Comment:
Does it make sense for margins to be negative like for drop-down animations?

Reply:
Shouldn't the client handle animations with a translation matrix on his side? I don't think we want users to use margins to animate things. Although, it might still be useful to have negative margins for other things.

Reply:
> I don't think we want users to use margins to animate things.

Why not? This would be such a cool feature for guake-style dropdowns and totally within the scope of the compositor.

Reply:
Negative margins sounds good to me